### PR TITLE
Cleaner environment

### PR DESCRIPTION
**Pull request by cjhahn**
Leak fewer transients into the environment, make the script set -u clean, handle shell metachars in paths, and make fatal errors fatal.

Several things still leak into the user environment, but those would be hard to stop or would trade documentation and didn't seem worth it.

Bear in mind I don't have a setup with which to test this, so there may be bugs.

### CODE REVIEW

File: postactivate [hunk 1]
```diff
@@ -1,122 +1,91 @@
 #!/bin/bash
-# This hook is run after every virtualenv is activated.
+# This hook is run after every virtualenv is activated.  It sets a shell alias
+# ('manage', by default) to the django manage.py, exports an appropriate value
+# for DJANGO_SETTINGS_MODULE, and changes directory into the project's root.
```
Comment (jbisbee) [inline postactivate:4]: Thank you for taking the time to do this and the script description :+1: I've been in a bit of bash hell forcing myself to learn and figured if you took a look at it, I would learn quite a bit from your edits/comments.

File: postactivate [hunk 2]
```diff
@@ -1,122 +1,91 @@
...
-}
-
-hget() {
-    if [[ -e $DJANGO_VIRTUALENV_NAME_MAP ]] 
-    then
-        grep "^$1 " $DJANGO_VIRTUALENV_NAME_MAP | awk '{ print $2 };'
-    else
-        echo ""
-    fi
-}
```
Comment (cjhahn) [inline postactivate:31]: Lines 13 - 31 being removed here weren't there when I started, and are an alternate solution to my lines 13 - 19.

If you go with A1 P's solution, just quote the uses of $DJANGO_VIRTUALENV_NAME_MAP (and use the "function hput {" syntax to stay consistent ;)

File: postactivate [hunk 3]
```diff
@@ -1,122 +1,91 @@
...
-then
-    export DJANGO_MANAGE_PY_ALIAS="manage"
-fi
+declare DJANGO_MANAGE_PY_ALIAS="${DJANGO_MANAGE_PY_ALIAS:-manage}"
+
 
 function determine_django_settings_module {
-    dir=$1
-    user=$2
+    declare project_dir="${1:?argument missing}" user="${2?:argument missing}"
```
Comment (cjhahn) [inline postactivate:32]: Throughout, I've added 'declare' to variable declarations.  Think of it just like Perl's 'local'.

File: postactivate [hunk 4]
```diff
@@ -1,122 +1,91 @@
...
-    fi
-}
-
-virtual_env_name=$(basename $VIRTUAL_ENV)
-django_project_name=$(hget $virtual_env_name)
-if [[ -z $django_project_name ]]
-then
-    django_project_name=$virtual_env_name
+# set DJANGO_PROJECTS_SRC_DIR to parent directory of all your project sources.  I use $HOME/src
+: "${DJANGO_PROJECTS_SRC_DIR:?"not defined/exported.  Set to parent directory of all your project sources."}" \
```
Comment (cjhahn) [inline postactivate:10]: This is one of my favourite tricks:
the ':' builtin just takes every argument you give it and returns true.  Perfect when all you want is to return true, but also great when all you really want to do is take advantage of bash's creative parameter expansion.
Like the ${var:?msg} expansion, which expands to the value of var unless var is null or unset, in which case it prints msg onto stderr and exits non-interactive shells.

**Pull request closed without merging**